src/kbc_manager.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::error::Error;
+use std::sync::Mutex;
+use string_error::into_err;
+
+use super::kbc_modules;
+
+// KbcInterface is a standard interface that all KBC modules need to implement.
+pub trait KbcInterface {
+    fn check(&self) -> KbcCheckInfo;
+    fn decrypt_payload(&mut self, messages: &str) -> Result<Vec<u8>, Box<dyn Error>>;
+}
+
+// KbcCheckInfo is used by KBC module instances to report their internal status to AA.
+pub struct KbcCheckInfo {
+    pub kbs_info: Vec<String>,
+    // In the future, more KBC status fields will be expanded here.
+}
+
+lazy_static! {
+    pub static ref KBC_MANAGER: Mutex<KbcManager> = Mutex::new(KbcManager::new());
+}
+
+pub struct KbcManager {
+    kbc_instance_list: HashMap<String, Box<dyn KbcInterface + Sync + Send>>,
+}
+
+impl KbcManager {
+    pub fn new() -> KbcManager {
+        KbcManager {
+            kbc_instance_list: HashMap::new(),
+        }
+    }
+
+    // This method is called when AA starts.
+    // It will create and instantiate all KBC modules supported by AA,
+    // and register them into the kbc_instance_list of KBC_MANAGER.
+    // The list of KBC modules supported by AA is the feature default field specified at compile time,
+    // please refer to ../Cargo.toml for details
+    pub fn init(&mut self) {
+        #[cfg(feature = "sample_kbc")]
+        {
+            let sample_kbc_instance = Box::new(kbc_modules::sample_kbc::SampleKbc::new());
+            self.kbc_instance_list
+                .insert(String::from("sample_kbc"), sample_kbc_instance);
+        }
+        ()
+    }
+
+    pub fn call_kbc_decrypt(
+        &mut self,
+        kbc_name: &str,
+        messages: &str,
+    ) -> Result<Vec<u8>, Box<dyn Error>> {
+        let kbc_instance = match self.kbc_instance_list.get_mut(kbc_name) {
```

Comment:
I guess this allows us to use a different KBC for each wrapped key. Will we ever actually do this? Could we eliminate the KBC Manager entirely by only initializing one KBC? We will probably want to reuse encrypted container images with different KBCs. Do we want to have the KBC specified as part of the image? Another approach would be to have the AA detect which platform it is running on and choose the KBC appropriately.

Reply:
We set the KBC instance created after AA runs as the compile time option. You can either compile and support only one KBC, or specify several different kbcs at compile time.  AA will create and instanciate each and every KBC module it supports (The supported list is an build time option) when it runs. The first init call should be triggered by the first corresponding ocicrypt-rs request AA gets for a given KBC.. This solution can flexibly support various situations, and hand over the type control of KBC to the compiler (user) of AA.

Reply:
I think there are some issues. Most importantly, I'm not sure it's a good idea to bake the KBC into the container image via a layer annotation. That means that a container image can only be used with one KBC, which most likely means you need different container images to run on different platforms. That's not ideal. Am I missing something?

Less of an issue, but a possible simplification, is to only support one KBC per AA. That would eliminate the need for the KBC manager. Why do we need to support multiple at once?

To account for both of these things, you could automatically select one KBC at startup and use that to fulfill all of the keyunwrap requests. Basically the AA would recognize that it's on a certain platform and use the correct KBC for that platform. This could be configurable, of course.

Reply:
The information "which KBC to select" is generated by ocicrypt and embedded in the DC field of ocicrypt AA protocol; The KBS information (such as IP address, URL, etc.) is written in the annotation field of the container image and specified by the user encrypting the image.

On the other hand, AA supports different kbcs not to support different hw-tee platforms, but to support different kbc-kbs communication protocols (such as isecl, GOP, etc.). Since AA cannot predict which KBS is specified in the annotation in the container image，it is necessary to instantiate the KBC corresponding to all possible KBS at runtime of AA (The KBC‘s supported list is an build time option). FYI: @hdxia

Reply:
But should the KBC information really be a property of the container image? I think it should be a property of the pod. The annotation for each layer could contain the key id, but the information about which KBC to use, the KBC URL, etc. would be in a Kubernetes annotation for the pod. As long as you provide a KBC to the pod that can provide a key matching the key id you can decrypt the container. Images shouldn't have a dependency on using a particular KBC.

Reply:
Please see the figure below. 
![截屏2021-09-06 上午10 52 10](https://user-images.githubusercontent.com/35883144/132153976-98c8818c-d2d0-40ba-a796-6969691948bf.png)

The user encrypts his container image through the key broker service (KBS), records the KBS and encryption operation information (eg: URL, key ID, etc.) in the annotation field of the container image, and then uploads the image to the public network. After the kata agent performs the pull image operation, it needs to ask the attention agent for help to decrypt the image. At this time, the attention agent will parse the content in the annotation sent by the kata agent, access the key broker service according to the KBS information (URL) recorded therein, finally obtain the KEK, and decrypt the container image layer.

Therefore, KBS URL, key ID and other information need to be recorded in the annotation field of the container image, while KBC information does not need to exist in the container image. KBC information is a kubernetes annotation for the pod. @jiazhang0

Reply:
Thanks for the comment. I think we may have some deviations in some terms. KBS: remote key broker service, KBC: AA's KBC module. KBS information needs to exist in the container image, but KBC information will not be stored in the container image

Reply:
@fitzthum here are ours answers about your questions.
1. For quesion: Could we eliminate the KBC Manager entirely by only initializing one KBC?
We think AA needs this KBC manager to support differenct type KBS platforms. As exapmle:
Intel SecL KBS (which supports TDX and SGX simultaneously):
It uses common HW-TEE remote attestation protocol to: 
  1>verify the HW-TEE's identity
  2>exchange a secret key between KBS and KBC
So a trusted and security channel can be built after Remote attestation finished. After that ISecL's specific KBC will send a getKey command to BKS to get corresponding key. This getKey command is ISecL's private protocol command.

Alibaba's EAA KBS (which supports TDX, SGX and SEV simultaneously):
It uses common HW-TEE remote attestation protocol to build a confidential and security TLS channel between KBS and KBC. This special TLS technology is based on Intel Lab's paper, the address is: https://arxiv.org/ftp/arxiv/papers/1801/1801.05863.pdf  (If you have any interesting about this).
After this TLS channel is built, EAA KBC can send getKey, unwrapKey .... commands to KBS to execute some expected operations.
These getKey, unwrapKey commands are all EAA KBS's private commands.

The upper two KBSs all use a common remote attestation protocol to check HW-TEE's identity and build a trust and security channel.
After that, it will use this channel to send KBS's private protocol commands.

So we think we can't use a single remote attestation protocol supported KBC instance, because this KBC can't support potential KBS's private protocol.


2. For question: Do we want to have the KBC specified as part of the image?
We think it's needed. 
The reason is there maybe new KBSs which supports SGX/TDX/SEV simultaneously will be developed in future. If AA can't get exact KBC's name info,  it can't choose exactly KBC.

3. For suggestion: Another approach would be to have the AA detect which platform it is running on and choose the KBC appropriately.
We think maybe we can't use this approach, becasue there maybe exist different KBSs which support SGX/TDX/SEV simultaneously. So only the platform type info isn't enough to choose the correctly KBC.

Reply:
Ok, forget about automatic detection of the KBC for now. I meant something slightly different by that which I will explain later. There are really two separate questions. 

1. Are images portable? In your diagram the keyid and the KBS URL are part of the image annotation. Obviously the key id has to be part of the annotation because it is specific to the layer or at least could be. The question is whether we need to have KBS-specific information (like the URL) as part of the annotation. If we do, then the image would seemingly only be usable with one KBS. It is still a bit unclear where the KBC information enters the picture, but let's just focus on the KBS URL for now. Could the KBS URL be a parameter of the pod rather than the image? This would mean that each pod would have to use the same KBS, but that you could switch KBS (either to a different type of KBS or to a similar one at a different location) at any time without updating the image. The KBS URL would be a Kubernetes annotation for the pod. You could point a pod to any KBS and as long as that KBS was able to provide a key given a keyid, you could decrypt the image.

2. Do we need the KBC manager? This is a lower priority. The question is whether we need more than one KBC per pod. Obviously we will need to support more than one KBC overall because there will be different protocols. The KBC manager is basically a level of indirection that allows us to instantiate multiple KBCs and then choose the correct one for each request. If the KBS were a parameter of the pod, we would only have one KBC per pod and we seemingly would not need the KBC manager. Even if we have the KBS URL in the annotation, are users really going to need different KBCs in one pod? If not, it seems like we could simplify by just allowing them to configure the pod to use one KBC which would field every request. (As an aside, what I was thinking with automatic detection is that the user would specify one KBC or each platform in the pod config and then the AA would detect the platform and use the corresponding KBC). 

What is the status of the EAA KBS? Is it open source? Is it going to be?

Reply:
Hi @fitzthum,
OK, let's align whether these KBS URL should be set as pod's configuration firstly.
I think it's workable to set URL to pod. And these info maybe can be passed to AA via key provider protocol.
But put these KBS URL info into Pod's configuation file will make this design unscalable. As example, if the image's each layer was encrypted by different user via different KBS, and each layer's encrypted key is also different. Our KBC manager design want to handle this complicated scenario. Following is our design's arch:
![image](https://user-images.githubusercontent.com/83809742/132441385-5def73bb-eb3f-444c-9d4d-331007d9f878.png)
In order to support this scenario, each pod need several KBCs simtaneously.

Yes, the EAA KBS is open source. And it's been developing now. It's code is in inclavare-containers project: https://github.com/alibaba/inclavare-containers

Reply:
I agree with @fitzthum that things would be more portable when the KBS URL comes from the pod, i.e. this is something the kata agent should get from the host and then pass that to the AA through e.g. the key unwrap parameters. Longer term we could decide we want to support a model where if the image layers come with KBS URLs, then this gets precedence over the pod specification. Then the image non portability becomes an image owner decision. But first I think we should assume the KBS URL(s) should come from the pod spec.

Now, on the KBC modularity and detection, I first have a question for you @voidC1326 and @liangzhou121: Let's say AA is trying to get a key and got the KBS URL for that key from ocicrypt. Then how does it know which KBC to use for this key? In other words, how does it map a KBS URL (The only information it has) to a KBC? How would the AA map e.g. `https://my-cloud.com/tenant-foo/kbs` to an isecl or eaa KBC?

Reply:
Hi @sameo 
- Is it possable to get KBC selection info also from pod's configuration? Like  **iscel**::https://my-cloud.com/tenant-foo/kbs ? Because users know the KBS address they used, so suppose they also know the KBS type.
- Or if AA can't get any KBC info from pod, it maybe need to poll and try all supported KBC module until it can get decryption key's **Kid**'s corresponding key successfully by one KBC.

Reply:
Hi @liangzhou121!

We could get it from a separate pod annotation, yes. I would not stuff it into the KBS URL annotation though. I wonder what are the security implications though, as the pod spec is controlled by the host. There may be cases where the host could use that information to try and establish secure links between a pod and a KBS that would support multiple protocols for example.

In your diagram you seem to imply that this information could also come from the agent config file. Were you thinking of mapping URLs/hosts/IPs to KBC names? The config file being controlled by the guest owner, and part of the TCB, I feel this is a safer option.

Reply:
@fitzthum 

>  If the KBS were a parameter of the pod, we would only have one KBC per pod and we seemingly would not need the KBC manager

A use case that may not sound too exotic/complex: Building a guest image that could support different key broker protocols. I'm a cloud provider and I want to provide my customers with a measured guest image that could run against e.g. isecl or eaa, depending on where they choose to host their encrypted container images. I think in that case we'd need a KBC manager and allow for dynamically switching between protocols, depending on how a pod expects to talk to its KBS. There would only be one KBC per pod, but different pods could use different KBCs.
Does that make sense? Or are we over engineering here?

Reply:
@liangzhou121

> I think it's workable to set URL to pod. And these info maybe can be passed to AA via key provider protocol or some else way. Do you have any suggestion about how to pass these info to AA?

One option is to set the KBS URL in a Kubernetes annotation. This would probably end up being part of the kernel commandline. The URL could also be part of the agent config file. Both of these are measured, which may or may not be a good thing. On the one hand, including the KBS URL in the launch measurement guarantees that it is correct. On the other hand, it increases the burden for changing the KBS URL (guest owner must recalculate measurement). It's not totally clear if the KBS URL needs to be measured for security purposes. If a malicious host changes the KBS URL the images won't decrypt.
> 
> Secondly, whether need to support more than one KBC in per pod.
> Yes, we think it's needed because it will enhance this design's scalability. As example, if the image's each layer was encrypted by different user via different KBS, and each layer's encrypted key is also different.

This is an interesting use-case. The KBC manager does make things more flexible and it isn't super complex, so I'm not totally against it. I'm not sure how common it would be for people to actually use images like this. Ultimately the layers are all being decrypted by the same user, so the two KBSes have some level of mutual trust. Maybe the keys could be consolidated to one KBS?

This approach would cause some issues with the one-URL-per-pod approach. If different layers are going to use different KBSes, different layers might also need different KBS URLs. That takes us back to the container images not being very portable.

Reply:
Hi @sameo @fitzthum 
We agree on your suggestions about KBS URLs and KBC names passed from Pod annotation or kata-agent config file. And we suggest to open another issue to discuss the details about their provenances.

Here is the summarized process about how to pass KBS Urls and KBC names:
- The caller of AA is responsible for passing a specific KBS url and KBC name to AA via key provider protocol.
- After receiving them, KBC manager will dynamically select and load the specified KBC module instance (So it will align with per-KBC-per-pod/per-URL-per-pod conception).

If there is no objection, we will document well about these details in IMPLEMENTATION.md in next PR. Are there any pending issues to be addressed by this PR?

Reply:
> What I mean by not having a KBC manager is really just that we wouldn't parse every unwrap request to dynamically switch between KBCs. Instead the AA would instantiate one KBC on startup and then use that to directly service every request. This would still give us lots of options for how the AA selects which KBC to use.
> 

I see. So what you're proposing would avoid having to match every request to a KBC, by picking one KBC for the whole pod lifetime. How the KBC is selected is another story.
Assuming that all container images for a given pod would use the same protocol/KBC, it does make sense to me. And I think that for now, the one single protocol/KBC per pod is a safe assumption. Until someone comes with a compelling, real use case for supporting different protocols/KBC per either container or per image layers, then I believe we should start with the simpler 1:1 protocol to pod mapping assumption. 

> I briefly mentioned one scheme that the AA could use to select KBC automatically. Basically, the user could specify one KBC per platform (TDX, SEV, etc.). Then the AA could detect the platform and use the preferred KBC. This would allow an unmodified guest image to work across platforms (although maybe not a generic one built by the CSP). The guest image would need to be built with each of the KBCs specified in the config, which isn't super efficient. Since some KBSes work with multiple platforms, maybe the platform isn't the biggest concern.
>

Given that we already have one example (eaa) that plans to support both TDX and SEV, mapping a platform to a KBC would make the guest image specific to the guest owner, but still reusable across platforms. 

>  It seems like the question is really how the AA would figure out which protocol it should use. 

Or maybe in most cases it does not need to figure it out and the information comes from the kata-agent/ocicrypt, as part of the key unwrap request parameters? I think we're looking at several options:

1. Pass the selected KBC as a pod annotation. The kata-agent would get that information and then pass it down to the AA.
2. Use a guest owner defined and measured mapping between platforms (TDX, SEV, etc) and KBC. This could be part of the kata-agent configuration file and would leave it to the kata-agent to detect the hardware platform and then select a KBC.
3. Use a guest owner defined and measured mapping between KBS hosts and KBC. This is fairly similar to the previous option, in the sense that the kata-agent would parse that information and select a KBC to be passed down to the AA.

In all 3 cases, it seems to me that the AA should only rely on what it's getting through the key provider interface, and that includes the KBC. Wether the AA decides to hard wire one KBC after the first request and only support that or if we want it to dynamically pick one with each request is almost an implementation detail. Based on the one KBC per pod assumption, I would go with instanciating one KBC for the whole pod lifetime. It's simpler and more efficient imho.

> Is it going to have to reach out to the KBS URL and negotiate the protocol?

That would be ideal, although that means this URL must then be trusted.

> Speaking of the KBS URL, if the CSP provides a generic image, where will the KBS URL be specified. If we want to measure the KBS URL, it might be hard to reconcile that with having a generic image. Do we need to measure the KBS URL?

If the KBS URL comes from a pod annotation and the host tampers with it, what could happen? The AA talks to a rogue KBS, sends a quote and could even get some rogue keys back from it. But the image layers decryption or verification would then fail, iiuc. Maybe I'm missing something here, does it make sense to you?

Reply:
@sameo 

Maybe we should just start from option 1 given that it is the simplest. 

One thing to note is that not every KBC will actually need a KBS URL. For instance a KBC for SEV(-ES) might rely only on boot-time secret injection to get secrets. Maybe the KBS URL should be part of a more general KBC option field. Some KBCs could be passed a filesystem path where the keys are baked into the guest image instead of a KBS URL.

I think the KBS URL probably does not need to be measured. What is the mechanism for conveying the URL from the Kubernetes annotation to the Kata Agent? Will it pass through the kernel commandline (which is measured by SEV)?

Reply:
> One thing to note is that not every KBC will actually need a KBS URL. For instance a KBC for SEV(-ES) might rely only on boot-time secret injection to get secrets. Maybe the KBS URL should be part of a more general KBC option field. Some KBCs could be passed a filesystem path where the keys are baked into the guest image instead of a KBS URL.

I think URL have already been a general name definitation and prefer don't merge into KBC option field. Because we only have 1:1 URL and KBC name mapping.
About how the URL will be parsed and used, it should be each KBC instance's specific implementation. AA's KBC manager will pass this URL parameter to KBC instance directly during KBC module's instaniation. So it should have no impact on AA's current design.

Reply:
Hi @sameo @fitzthum 
As we have already reached agreement of following AA's design:
- AA receives KBS Url and KBC name from key provider protocol
- AA's KBC manager (act as KBC selector + KBC loader)  will load only one KBC instance per pod and go with this instance for the whole pod lifetime.
- per KBC instance per pod && per KBS url per pod.

So I opened an separately new issue to diccuss the how to configure and store KBS Url and KBC name mapping dedicatedly.
https://github.com/containers/attestation-agent/issues/6

Reply:
Hi @sameo @fitzthum 
After our discussion above, I summarize the latest AA architecture and operation process as follows:

![截屏2021-09-10 上午11 44 16](https://user-images.githubusercontent.com/35883144/132796486-86a0f24a-b31d-4012-b001-4bbaca9cb368.png)

In the above figure, how to pass KBC name and KBS URL to Kata agent is still represented as a question mark. We have opened a new issue (#6) for discussion.

Reply:
> @sameo
> 
> Maybe we should just start from option 1 given that it is the simplest.

I agree.

> One thing to note is that not every KBC will actually need a KBS URL. For instance a KBC for SEV(-ES) might rely only on boot-time secret injection to get secrets. Maybe the KBS URL should be part of a more general KBC option field. Some KBCs could be passed a filesystem path where the keys are baked into the guest image instead of a KBS URL.
> 

Or we could say that we're passing a URI, where the scheme could be `http` or `file`, and always have a `(KBC, KBS URI)` pair.

> I think the KBS URL probably does not need to be measured. What is the mechanism for conveying the URL from the Kubernetes annotation to the Kata Agent? Will it pass through the kernel commandline (which is measured by SEV)?

Right now, we don't forward any pod annotation to the agent, we only carry the OCI spec ones for each container (if there are any). I think we may want to extend the [Sandbox creation request](https://github.com/kata-containers/kata-containers/blob/main/src/agent/protocols/protos/agent.proto#L276) payload to include an optional KBC+KBS URI pair.

Reply:
> Hi @sameo @fitzthum
> As we have already reached agreement of following AA's design:
> 
>     * AA receives KBS Url and KBC name from key provider protocol
> 
>     * AA's KBC manager (act as KBC selector + KBC loader)  will load only one KBC instance per pod and go with this instance for the whole pod lifetime.
> 
>     * per KBC instance per pod && per KBS url per pod.
> 

That makes sense to me. @fitzthum are you ok with that approach?

Reply:
> > Hi @sameo @fitzthum
> > As we have already reached agreement of following AA's design:
> > ```
> > * AA receives KBS Url and KBC name from key provider protocol
> > 
> > * AA's KBC manager (act as KBC selector + KBC loader)  will load only one KBC instance per pod and go with this instance for the whole pod lifetime.
> > 
> > * per KBC instance per pod && per KBS url per pod.
> > ```
> 
> That makes sense to me. @fitzthum are you ok with that approach?

Yes this LGTM. The only minor thing is that for point 1, I'm not sure we need to extend the key provider protocol to pass the KBC name and KBS URI. The kata agent can probably just give the values to the AA when the Kata Agent starts the AA.

Reply:
I think KBC name and KBS URL should be transmitted through keyprovider protocol for the following two reasons:

1. In fact, we do not need to extend the current keyprovider protocol. The (KBC name, KBS URL) pair can be passed through the customize parameter fields reserved in the current keyprovider protocol.

2. Passing KBC name and KBS URL through keyprovider protocol instead of AA startup parameters can simplify AA architecture and startup process, unify AA parameter transmission interface, and facilitate future maintenance.

@fitzthum Do you agree to this ?

Reply:
@voidC1326
> 
> @fitzthum Do you agree to this ?

Not really a big deal either way. It seems like it would be a bit more direct to go from the Kata Agent to the AA (although maybe ocicrypt-rs will actually be the one starting the AA; not sure). If we pass the KBC name and KBS URI through the keyprovider protocol do we have to wait for the first unwrap request before we can setup the KBC? I'm sure it will work either way.

Reply:
> Not really a big deal either way. It seems like it would be a bit more direct to go from the Kata Agent to the AA (although maybe ocicrypt-rs will actually be the one starting the AA; not sure). If we pass the KBC name and KBS URI through the keyprovider protocol do we have to wait for the first unwrap request before we can setup the KBC? I'm sure it will work either way.
Yes, the KBC will be setup until the first unwrap request received. And I think this mechanism's performance impact to the first unwrap request can be ignored beacuse the KBC instance setup is simple. 
So we prefer to use key provider protocol to get KBS Url and KBC name becasue as voidC1326 mentioned it will make AA's arch simple.

Reply:
Hi @fitzthum @sameo 
I have already updated the PR and written the results of our discussion and the current implementation description of AA into the IMPLEMENTATION.md doc. Please take a moment to review them, thanks!

Reply:
I'll take a closer look tomorrow. One thing now is that we might want to switch KBS_URL to KBS_URI. @sameo and I discussed this a bit above, but basically some KBCs might read from a file rather than connecting over the network.